### PR TITLE
tlsfwd: inital commit

### PR DESCRIPTION
**Pull request by karasz**

### CODE REVIEW

**Comment by amery** [comment]
@karasz why doesn't it use `shared/proxy.Forward` ?

**Comment by karasz** [comment]
> @karasz why doesn't it use `shared/proxy.Forward` ?

no special reason just that I see them as separate

**Comment by amery** [comment]
> > @karasz why doesn't it use `shared/proxy.Forward` ?
> 
> no special reason just that I see them as separate

from here it looks like unjustified duplication :man_shrugging:

**Comment by karasz** [comment]
> from here it looks like unjustified duplication man_shrugging
it is.

**Comment by karasz** [comment]
> from here it looks like unjustified duplication man_shrugging

I am not sure I will manage to implement it today.

**Comment by karasz** [comment]
> @karasz why doesn't it use `shared/proxy.Forward` ?

@amery pushed another version that uses `shared/proxy.Forward` but I do not like the upstream IP picking that much

File: agent/tlsfwd/tlsfwd.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+	default:
+	}
+
+	upstream, err := addrFromName(c.ServerName)
+	if err != nil {
+		return err
+	}
+
+	errm := proxy.Forward(ctx, conn, upstream)
+	fmt.Println(errm)
```
Comment (amery) [inline agent/tlsfwd/tlsfwd.go:90]: please use our logger for debug prints

File: agent/tlsfwd/tlsfwd.go [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+	}
+}
+
+func tlsHandler(ctx context.Context, c *tls.ClientHelloInfo,
+	conn net.Conn,
+) error {
+	defer conn.Close()
+
+	select {
+	case <-ctx.Done():
```
Comment (amery) [inline agent/tlsfwd/tlsfwd.go:79]: this ctx.Done() test should be done by the resolver

File: agent/tlsfwd/tlsfwd.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+func addrFromName(s string) (netip.AddrPort, error) {
+	upIPs, err := net.LookupIP(s)
+	if err != nil {
+		fmt.Println(err)
+		return netip.AddrPort{}, err
+	}
+
+	addrToUse, ok := netip.AddrFromSlice(upIPs[0])
```
Comment (amery) [inline agent/tlsfwd/tlsfwd.go:101]: please use `core`

File: agent/tlsfwd/tlsfwd.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+	defer func() {
+		if !ok {
+			for _, k := range s.ds {
+				_ = k.Close
+			}
+		}
+	}()
+
+	errHTTP := s.prepareRedir()
```
Comment (amery) [inline agent/tlsfwd/tlsfwd.go:123]: I see no value in making a unique err name here

File: agent/tlsfwd/tlsfwd.go [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+		return err
+	}
+	ok = true
+	return s.wg.Wait()
+}
+
+func (s *Server) prepareRedir() error {
+	if s.redir {
+		s.wg.Go(func() error {
+			return http.ListenAndServe(":80",
```
Comment (amery) [inline agent/tlsfwd/tlsfwd.go:139]: please make both ports configuration parameters

File: agent/tlsfwd/tlsfwd.go [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+	return nil
+}
+
+func (s *Server) prepareDispatchers() error {
+	for _, p := range s.ps {
+		dispatcher := &sni.Dispatcher{
+			GetHandler: getDefaultTLSHandler,
+		}
+		s.ds = append(s.ds, dispatcher)
+		ll, err := net.Listen("tcp", p)
```
Comment (amery) [inline agent/tlsfwd/tlsfwd.go:152]: for better error handling all listeners should be opened before spawning the workers

File: agent/tlsfwd/tlsfwd.go [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+	return s.Err()
+}
+
+func (s *Server) tryCancel(err error) {
+	// once
+	if s.cancelled.CompareAndSwap(false, true) {
+		const msg = "Initiating shutdown"
+
+		if err != nil {
+			fmt.Printf("%s: %s", msg, err.Error())
```
Comment (amery) [inline agent/tlsfwd/tlsfwd.go:183]: please use our logger

File: cmd/tlsfwd/main.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+	if *usRedirect {
+		s, err = tlsfwd.NewTLSFwd(ports.Ports, tlsfwd.SetRedir)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+	s.Serve()
+}
+
+func unique(intSlice []int) []int {
```
Comment (amery) [inline cmd/tlsfwd/main.go:61]: a `SliceUnique[T comparable]` helper in `core`?

**Review by amery — CHANGES_REQUESTED** [review]

**Comment by amery** [comment]
also, please rebase it on top of the current `master` to avoid conflicts

File: agent/tlsfwd/tlsfwd.go (hunk 4 above)
Comment (karasz) [reply agent/tlsfwd/tlsfwd.go]: fixed

**Comment by karasz** [comment]
> also, please rebase it on top of the current `master` to avoid conflicts

rebased

File: agent/tlsfwd/tlsfwd.go (hunk 5 above)
Comment (karasz) [reply agent/tlsfwd/tlsfwd.go]: fixed

File: agent/tlsfwd/tlsfwd.go (hunk 3 above)
Comment (karasz) [reply agent/tlsfwd/tlsfwd.go:105]: I don't think we have in `core` something that transforms a `net.IP` to a `netip.Addr`

File: agent/tlsfwd/tlsfwd.go (hunk 6 above)
Comment (karasz) [reply agent/tlsfwd/tlsfwd.go:153]: pushed

File: agent/tlsfwd/tlsfwd.go (hunk 7 above)
Comment (karasz) [reply agent/tlsfwd/tlsfwd.go:188]: I don't want a logger, tlsfwd is not a proxy, it is just a convenience one shot thing to test stuff

**Review by amery — CHANGES_REQUESTED** [review]

**Comment by karasz** [comment]
need to do more thinking.

**Pull request closed without merging**